### PR TITLE
Elastic: Fix parsing for millisecond number timestamps

### PR DESCRIPTION
**Pull request by aocenas**
While trying to fix fake-data-gen for Elastic I found issue with how we parse the response.

In fake-data-gen timestamp field is number of milliseconds so a number. This should be fine as it should not be a problem to convert to Date but we also parse the `fields` from the Elastic response and there the time field is a `string[]`. So there was a code to handle that it is an array but not that there is a milliseconds number that is string encoded.

The `fields` attribute is there because we explicitly ask for it when doing document query but as timestamp is always inside the `_source` and in correct format this seems to be unnecessary and was removed and the timestamp is only parsed from `_source`.

### CODE REVIEW

**Comment by aocenas** [comment]
@kaydelaney @marefr Can you check this? I think you two wrote the original code so just wan't to make sure I am correct with the `fields` parsing and am not missing something.

**Comment by davkal** [comment]
Timestamp in v5.6 seem to be integers, so this PR breaks older versions.

Screenshot 2019-11-11 at 15 47 51

**Comment by davkal** [comment]
See #20304 for a working docker block for v5.

**Comment by aocenas** [comment]
@davkal 
> Timestamp in v5.6 seem to be integers, so this PR breaks older versions.

Need to check older versions but this change should explicitly make integers as timestamps work as they do not work in master now. Also this should only be a problem for logs not metrics, because with logs the timestamp field is duplicated and would be once in `fields` as an array of strings (even though it is an integer it is stringified) and in `_source` where it is properly an integer.

**Comment by aocenas** [comment]
Here is the response from the fake data logs:
Screenshot 2019-11-11 at 16 47 44

As you can see `@timestamp` is string in the `fields` while integer in `_source`. This results in:
Screenshot 2019-11-11 at 16 47 51

File: public/app/plugins/datasource/elasticsearch/elastic_response.ts [hunk 1]
```diff
@@ -424,40 +424,26 @@ export class ElasticResponse {
...
         const flattened = hit._source ? flatten(hit._source, null) : {};
-        doc = {};
-        doc[this.targets[0].timeField] = null;
-        doc = {
-          ...doc,
+        const doc = {
           _id: hit._id,
           _type: hit._type,
           _index: hit._index,
+          _source: { ...flattened },
```
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/elastic_response.ts]: `flattened` is now both in doc root and in `_source` prop?

**Review by marefr — COMMENTED** [review]
Would like to see some tests to understand what problem you're trying to solve.

The idea when implementing this was to use the same logic as when processing hits which is mainly (only) used in table panel:  https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/elasticsearch/elastic_response.ts#L315-L348.

File: public/app/plugins/datasource/elasticsearch/elastic_response.ts (hunk 1 above)
Comment (aocenas) [reply public/app/plugins/datasource/elasticsearch/elastic_response.ts]: it is the same as before, the code is just cleaned up a little

**Comment by aocenas** [comment]
@marefr Will add tests if this should be merged but I described the issue in previous comment, the `@timestamp` just isn't parsed correctly if it is stored as integer milliseconds in elastic (which for example is how fake-data-gen creates the logs).

File: public/app/plugins/datasource/elasticsearch/elastic_response.ts (hunk 1 above)
Comment (aocenas) [reply public/app/plugins/datasource/elasticsearch/elastic_response.ts]: It is the same as before just cleaned up a little. Before it was assigned to `_source` here https://github.com/grafana/grafana/pull/20290/files#diff-b2fa433b72195826bde52895d369871fL459

Comment (marefr) [reply public/app/plugins/datasource/elasticsearch/elastic_response.ts]: Sorry, my bad didn't notice that.

**Review by torkelo — COMMENTED** [review]
Think this should be fine, what versions of Elastic did you test with?

**Comment by marefr** [comment]
@aocenas I was unsure what parsing didn't work, only in logs panel or also in table panel. I guess this is related to parsing of timestamps for the logs panel?

**Comment by aocenas** [comment]
@marefr Yes this is when the query has `isLogQuery` as I said the issue is we then get timestamp column where value is a string something like `1234356546` and that is later not parseable by `datetime`.

**Comment by marefr** [comment]
For the table panel the date formatter is implemented like this:
https://github.com/grafana/grafana/blob/e9668fd2519af3191284d61d9214b137a6189f10/public/app/plugins/panel/table/renderer.ts#L109-L120

Can you use the same here maybe?

**Comment by aocenas** [comment]
@marefr tbh not sure why that would be better. The timestamp isn't string in elastic and is sent as number. The problem is we are not using the timestamp from `_source` which I think we should, but from `fields` attribute of the response where it is string. That just does not make sense to me, we should just use what is inside the `_source`. The `fields` if I understand correctly is there just because of the separate aggregation so we should not care about it here. I just wanted to confirm this.

**Comment by marefr** [comment]
@aocenas was just suggesting to find a way to handle this in same way since data frames will be used in table panel as well eventually and probably same kind of response for logs and table data. Not optimal to me to have different logic in multiple places. 

But I don't see any problems with your suggested changes if it solves your problems.

**Comment by torkelo** [comment]
The value formatting for the timestamp under _source could be different (I think) while under fields it should be what we ask for. 

When querying ES 5 raw documents the timestamp is an array of numbers. Not sure why it is an array :) 

```json
"fields": {
              "@timestamp": [
                1574090428170
              ]
            },
```

But not sure, maybe using it from _source should be always the same.

File: packages/grafana-data/src/dataframe/MutableDataFrame.ts [hunk 2]
```diff
@@ -231,11 +231,6 @@ export class MutableDataFrame<T = any> implements DataFrame, MutableVector<T> {
           field.parse = makeFieldParser(val, field);
         }
         val = field.parse(val);
-      } else if (field.type === FieldType.time && isArray(val)) {
-        if (!field.parse) {
-          field.parse = (val: any[]) => val[0] || undefined;
```
Comment (aocenas) [inline packages/grafana-data/src/dataframe/MutableDataFrame.ts]: @ryantxu can you check whether this automatic unboxing is needed somewhere else I did not find anything so far.

File: public/app/plugins/datasource/elasticsearch/specs/query_builder.test.ts [hunk 3]
```diff
@@ -1,573 +1,507 @@
...
-      bucketAggs: [
-        { type: 'terms', field: '@host', id: '2' },
-        { type: 'date_histogram', field: '@timestamp', id: '3' },
-      ],
-    });
-
-    expect(query.aggs['2'].terms.field).toBe('@host');
-    expect(query.aggs['2'].aggs['3'].date_histogram.field).toBe('@timestamp');
-  });
+  const builder = new ElasticQueryBuilder({ timeField: '@timestamp', esVersion: 2 });
```
Comment (aocenas) [inline public/app/plugins/datasource/elasticsearch/specs/query_builder.test.ts]: Mainly a refactor to do the tests for each version.

**Comment by aocenas** [comment]
@torkelo or @marefr can you take a look again into this?

File: packages/grafana-data/src/dataframe/MutableDataFrame.ts (hunk 2 above)
Comment (torkelo) [reply packages/grafana-data/src/dataframe/MutableDataFrame.ts]: could not find any

**Review by torkelo — APPROVED** [review]
Looks good!

**Pull request merged**